HistogramData rollout: algorithms DivideMD to IntegratePeaksMD

samueljackson92 opened the pull request:
## Description of work.
This PR is part of refactoring effort to use the new interface of the HistogramData module.
See https://github.com/mantidproject/mantid/issues/17641 for details.

The following algorithms were refactored to use HistogramData:
```
Framework/MDAlgorithms/src/DivideMD.cpp
Framework/MDAlgorithms/src/EqualToMD.cpp
Framework/MDAlgorithms/src/FindPeaksMD.cpp
Framework/MDAlgorithms/src/GetSpiceDataRawCountsFromMD.cpp
Framework/MDAlgorithms/src/GreaterThanMD.cpp
Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp
Framework/MDAlgorithms/src/IntegrateFlux.cpp
Framework/MDAlgorithms/src/IntegratePeaksCWSD.cpp
Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp
Framework/MDAlgorithms/src/IntegratePeaksMD.cpp
```

Performance tests were added for: N/a

The following algorithms have improved performance: N/a

## To test.

Code review.


Fixes #18656 .
Internal change, no release notes (potential performance improvements will be added to release notes as part of the umbrella issue).

LamarMoore commented on Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:
```diff
@@ -471,20 +463,16 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
-          wsProfile2D->dataY(i)[j] = signal_fit[j];
-          wsProfile2D->dataE(i)[j] = std::sqrt(signal_fit[j]);
-        }
+        setSpectrum(wsProfile2D, i, signal_fit);
```
Covered in previous comments. Shouldn't need `setSpectrum` unless it returns a histogram.

LamarMoore commented on Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:
```diff
@@ -555,18 +543,21 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
-        wsFit2D->dataX(i) = x;
-        wsDiff2D->dataX(i) = x;
-        FunctionDomain1DVector domain(x);
+        const auto &x = wsProfile2D->x(i);
```
sharedX?

samueljackson92 replied:
👍

LamarMoore commented on Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:
```diff
@@ -555,18 +543,21 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
+        auto &fit2DY = wsFit2D->mutableY(i);
+        auto &diff2DY = wsDiff2D->mutableY(i);
+        auto funcValues = yy.toVector();
+        fit2DY = funcValues;
```
Why not use `Histogram` arithmetic? You could do something like 
```
wsFit2D->setSharedY(i, wsDiff2D->y(i));
wsFit2D->mutableY(i) -= yy;
```

or something similar.

samueljackson92 replied:
👍

samueljackson92 commented:
@LamarMoore Sorry for the slow response. Thanks for your comments. I'll hopefully get a chance to work through them individually this week.

samueljackson92 commented on Framework/MDAlgorithms/src/GetSpiceDataRawCountsFromMD.cpp:
```diff
@@ -499,17 +501,12 @@ MatrixWorkspace_sptr GetSpiceDataRawCountsFromMD::createOutputWorkspace(
-      dataE[i] = 1.;
-  }
+  outws->setPoints(0, vecX);
+  outws->setCounts(0, vecY);
```
Agreed about the setting method. But here this is returning `1.0` if the Y value is `< 1.0` are well as performing the square root? So calculating it automatically could change the algorithm's output?

samueljackson92 commented on Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp:
```diff
@@ -145,8 +145,8 @@ void IntegrateEllipsoids::qListFromHistoWS(Integrate3DEvents &integrator,
     // get tof and counts
-    const Mantid::MantidVec &xVals = wksp->readX(i);
-    const Mantid::MantidVec &yVals = wksp->readY(i);
+    const auto &xVals = wksp->x(i);
```
Agreed.

samueljackson92 commented on Framework/MDAlgorithms/src/IntegrateEllipsoids.cpp:
```diff
@@ -559,17 +552,8 @@ void IntegrateEllipsoids::exec() {
-          wsProfile2D2->dataY(2)[j] = principalaxis3[j];
-          wsProfile2D2->dataE(2)[j] = std::sqrt(principalaxis3[j]);
-        }
+        setOutputWorkspaceData(
```
Agreed. `HistogramData::LinearGenerator` was the droid I was looking for.

samueljackson92 commented on Framework/MDAlgorithms/src/IntegrateFlux.cpp:
```diff
@@ -113,13 +113,16 @@ IntegrateFlux::createOutputWorkspace(const API::MatrixWorkspace &inputWS,
   // spectrum
-  for (double x = xMin; ix != X.end(); ++ix, x += dx) {
-    *ix = x;
-  }
```
Agreed.

samueljackson92 commented on Framework/MDAlgorithms/src/IntegrateFlux.cpp:
```diff
@@ -229,12 +233,12 @@ void IntegrateFlux::integrateSpectraHistograms(
+    auto inY = inputWS.y(sp); // make a copy
 
     // if it's a distribution y's must be multiplied by the bin widths
     if (isDistribution) {
```
Agreed.

samueljackson92 commented on Framework/MDAlgorithms/src/IntegratePeaksMD.cpp:
```diff
@@ -455,21 +452,19 @@ void IntegratePeaksMD::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
+        background_total =
+            std::accumulate(y.begin(), y.begin() + peakMin, background_total);
+        background_total =
+            std::accumulate(y.begin() + peakMax, y.end(), background_total);
```
Will do.

samueljackson92 commented:
@LamarMoore I think that has addressed most of your comments. Only remaining thing is your first comment on `GetSpiceDataRawCountsFromMD.cpp`, but perhaps this is incorrect behaviour anyway and we should just compute E automatically as suggested.

LamarMoore commented on Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:
```diff
@@ -411,28 +414,21 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
         signal_fit.clear();
-        for (size_t j = 0; j < numSteps; j++)
-          signal_fit.push_back(0.0);
+        std::fill(signal_fit.begin(), signal_fit.end(), 0.0);
```
Minor comment. If you used `Counts` instead of `std::vector` for signal_fit, you could just use `signal_fit = 0;`

LamarMoore commented on Framework/MDAlgorithms/src/IntegratePeaksMD2.cpp:
```diff
@@ -555,17 +548,19 @@ void IntegratePeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
-        }
+        auto funcValues = yy.toVector();
+
+        wsFit2D->mutableY(i) = funcValues;
```
```cpp
wsFit2D->mutableY(i) = std::move(funcValues);
wsDiff2D->setSharedY(i, wsProfile2D->sharedY(i));
wsDiff2D->mutableY(i) -= wsFit2D->y(i);
```
??

Edit:
```cpp
wsFit2D->mutableY(i) = yy.toVector();//don't think you need std::move in this case
```

samueljackson92 replied:
I need to get more used to thinking about where I can use `std::move`

samueljackson92 commented on Framework/MDAlgorithms/src/GetSpiceDataRawCountsFromMD.cpp:
```diff
@@ -499,17 +502,11 @@ MatrixWorkspace_sptr GetSpiceDataRawCountsFromMD::createOutputWorkspace(
+  outws->setHistogram(0, Points(std::move(vecX)), Counts(std::move(vecY)));
+
+  auto &dataE = outws->mutableE(0);
+  std::transform(vecY.begin(), vecY.end(), dataE.begin(),
```
Yes. `remove_if` that the algorithm I was thinking of. And you're right about it breaking the E data anyway with the move. I guess the E value isn't actually covered by a test...

LamarMoore commented:
If @SimonHeybrock does not have any further comments :shipit:
